sihl/src/core_configuration.ml

```diff
@@ -60,8 +58,76 @@ let envs_to_kv envs =
          | key :: values -> key, String.concat "" values)
 ;;
 
-(* TODO [jerben] Consider memoizing this or any consumer functions *)
+(* .env file handling *)
+let root_path () =
+  match Sys.getenv_opt "ROOT_PATH" with
+  | None | Some "" ->
+    let markers = [ ".git"; ".hg"; ".svn"; ".bzr"; "_darcs" ] in
+    let rec find_markers path_els =
+      let path = String.concat "/" path_els in
+      if List.exists
+           (fun marker -> Sys.file_exists (path ^ "/" ^ marker))
+           markers
+      then (
+        (* Path found => Write it into the env var to "memoize" it *)
+        Unix.putenv "ROOT_PATH" path;
+        Some path)
+      else (
+        match path_els with
+        | [] -> None
+        | _ -> find_markers @@ CCList.take (List.length path_els - 1) path_els)
+    in
+    find_markers @@ String.split_on_char '/' (Unix.getcwd ())
+  | Some path -> Some path
+;;
+
+let env_files_path () =
+  match Sys.getenv_opt "ENV_FILES_PATH" with
+  | None | Some "" -> root_path ()
+  | Some path -> Some path
+;;
+
+let read_env_file () =
+  match env_files_path () with
+  | Some path ->
+    let is_test =
+      match Sys.getenv_opt "SIHL_ENV" with
+      | Some "test" -> true
+      | _ -> false
+    in
+    let filename = path ^ "/" ^ if is_test then ".env.test" else ".env" in
+    let exists = CCIO.File.exists filename in
```

Comment:
Since this function is exposed, it is not just used at startup, but can be used at any time. Would it not make sense to keep `Lwt` here then?

Reply:
I added a comment to not use this function to read env vars. We can not keep `Lwt` because then `read_string` has to also return `Lwt`.

Reply:
It kind of makes sense, since now `read_string` could also trigger reading the env files, so should it not also be `Lwt`?

Reply:
Yes it makes sense from the implementation point of view, but we have to allow callers in a non-Lwt context, such as setting up a middleware.

Reply:
Got it, thanls!